File: Docusnap/database/modules/test_text_normalise.js
#!/usr/bin/env node
'use strict';

/**
 * database/modules/test_text_normalise.js
 * ---------------------------------------
 * PARITY guard: the JS normaliser must produce byte-identical output to the Python
 * reference on the SHARED golden corpus (python_backend/tests/normalise_corpus.json).
 * If this fails, the renderer and backend normalisers have drifted.
 *
 *   node database/modules/test_text_normalise.js
 */

const fs   = require('fs');
const path = require('path');
const { normaliseForTokens, tokenise } = require('./text_normalise');

function check(label, cond) { console.log(`  ${cond ? 'OK ' : 'BAD'} ${label}`); return cond; }

const corpusPath = path.join(__dirname, '..', '..', 'python_backend', 'tests', 'normalise_corpus.json');
const CORPUS = JSON.parse(fs.readFileSync(corpusPath, 'utf-8'));

function eqArr(a, b) { return Array.isArray(a) && Array.isArray(b) && a.length === b.length && a.every((x, i) => x === b[i]); }

function main() {
  let f = 0;
  console.log('text_normalise.js matches the shared golden corpus (Py/JS parity)');
  for (const c of CORPUS) {
    f += !check(`norm ${JSON.stringify(c.in)} -> ${JSON.stringify(c.norm)}`, normaliseForTokens(c.in) === c.norm);
    f += !check(`tokens ${JSON.stringify(c.in)}`, eqArr(tokenise(c.in), c.tokens));
  }
  if (f) { console.log(`\n${f} FAILED — JS normaliser diverged from the Python golden corpus.`); process.exit(1); }
  console.log(`\nAll ${CORPUS.length} normaliser cases match (Py/JS parity holds).`);
  process.exit(0);
}

main();

File: Docusnap/python_backend/tests/test_text_normalise.py
#!/usr/bin/env python3
"""
tests/test_text_normalise.py
----------------------------
Stability/regression guard for the compare-time normaliser against the SHARED
golden corpus (normalise_corpus.json). The JS twin asserts the SAME corpus
(database/modules/test_text_normalise.js), so a mismatch on either side = drift.

    py -3.12 python_backend/tests/test_text_normalise.py
"""
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))
from extraction.text_normalise import normalise_for_tokens, tokenise  # noqa: E402

CORPUS = json.loads((Path(__file__).parent / "normalise_corpus.json").read_text(encoding="utf-8"))


def check(label, cond):
    print(f"  {'OK ' if cond else 'BAD'} {label}")
    return bool(cond)


def main():
    f = 0
    print("text_normalise matches the shared golden corpus")
    for case in CORPUS:
        got_n = normalise_for_tokens(case["in"])
        got_t = tokenise(case["in"])
        # ascii() escapes non-ASCII so the Windows cp1252 console never chokes.
        lbl = ascii(case["in"])
        f += not check(f"norm {lbl} -> {ascii(case['norm'])}", got_n == case["norm"])
        f += not check(f"tokens {lbl}", got_t == case["tokens"])
    if f:
        print(f"\n{f} FAILED — normaliser drifted from the golden corpus.")
        return 1
    print(f"\nAll {len(CORPUS)} normaliser cases match the golden corpus.")
    return 0


if __name__ == "__main__":
    sys.exit(main())
